File: Kotlin/MergeSort.kt
/**
 *  Algorithm: Merge Sorting
 *  Language: Kotlin
 *  Input:
 *   a) n: Size of Array
 *   b) arr: Integer Array of size n
 *  Output:
 *   a) Array before applying Merge Sorting
 *   b) Array after applying Merge Sorting
 *  Time Complexity: O(n * log n)
 *  Space Complexity: O(n)
 *
 *  Sample Input:
 *  Enter the size of the integer array:
 *  5
 *  Enter an integer array:
 *  5 4 8 2 1
 *
 *  Sample Output:
 *  Array before sorting is:
 *  5 4 8 2 1
 *  Array after sorting is:
 *  1 2 4 5 8
 */


import MergeSort.Companion.printIArray
import java.util.*

class MergeSort {
    fun mergeSort(array: IntArray) {
        val n = array.size
        if (n < 2) return
        val mid = n / 2
        val left = IntArray(mid)
        val right = IntArray(n - mid)
        for (i in 0 until mid) left[i] = array[i]
        for (i in mid until n) right[i - mid] = array[i]
        mergeSort(left)
        mergeSort(right)
        merge(left, right, array)
    }

    private fun merge(left: IntArray, right: IntArray, array: IntArray) {
        val nLeft = left.size
        val nRight = right.size
        var i = 0
        var j = 0
        var k = 0
        while (i < nLeft && j < nRight) {
            if (left[i] <= right[j]) {
                array[k] = left[i]
                k++
                i++
            } else {
                array[k] = right[j]
                k++
                j++
            }
        }
        while (i < nLeft) {
            array[k] = left[i]
            i++
            k++
        }
        while (j < nRight) {
            array[k] = right[j]
            j++
            k++
        }
    }

    companion object {
        fun printIArray(array: IntArray) {
            println(array.joinToString(" "))
        }
    }
}

fun main() {
    val scanner = Scanner(System.`in`)
    println("Enter the size of the integer array: ")
    val n = scanner.nextInt()
    println("Enter an integer array: ")
    val arr = IntArray(n)
    arr.indices.forEach {
        arr[it] = scanner.nextInt()
    }
    val merge = MergeSort()
    println("Array before sorting is: ")
    printIArray(arr)
    merge.mergeSort(arr)
    println("Array after sorting is: ")
    printIArray(arr)
}

File: JavaScript/mergesort.js
const unsortedArr = [
  31,
  27,
  28,
  42,
  13,
  8,
  11,
  30,
  17,
  41,
  15,
  43,
  1,
  36,
  9,
  16,
  20,
  35,
  48,
  37,
  7,
  26,
  34,
  21,
  22,
  6,
  29,
  32,
  49,
  10,
  12,
  19,
  24,
  38,
  5,
  14,
  44,
  40,
  3,
  50,
  46,
  25,
  18,
  33,
  47,
  4,
  45,
  39,
  23,
  2,
];

const merge = (arr1, arr2) => {
  let sorted = [];

  while (arr1.length && arr2.length) {
    if (arr1[0] < arr2[0]) sorted.push(arr1.shift());
    else sorted.push(arr2.shift());
  }

  return sorted.concat(arr1.slice().concat(arr2.slice()));
};

const mergeSort = (arr) => {
  if (arr.length <= 1) return arr;
  let mid = Math.floor(arr.length / 2),
    left = mergeSort(arr.slice(0, mid)),
    right = mergeSort(arr.slice(mid));

  return merge(left, right);
};

console.log(mergeSort(unsortedArr));

/**
 * Sample output
 * [
   1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,
  12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22,
  23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
  34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44,
  45, 46, 47, 48, 49, 50
]
 Time Complexity: O(n(log n))
 Space Complexity: O(n)
    
 */
